coins/monero/src/wallet/send/mod.rs

```diff
@@ -332,56 +422,89 @@ impl SignableTransaction {
-    let extra = Extra::fee_weight(outputs, true, has_payment_id, data.as_ref());
+    let extra = Extra::fee_weight(outputs, additional, has_payment_id, data.as_ref());
 
     // https://github.com/monero-project/monero/pull/8733
     const MAX_EXTRA_SIZE: usize = 1060;
     if extra > MAX_EXTRA_SIZE {
       Err(TransactionError::TooMuchData)?;
     }
 
-    // This is a extremely heavy fee weight estimation which can only be trusted for two things
-    // 1) Ensuring we have enough for whatever fee we end up using
-    // 2) Ensuring we aren't over the max size
-    let estimated_tx_size = Transaction::fee_weight(protocol, inputs.len(), outputs, extra);
+    // Caclculate weight of decoys
+    let decoy_weights =
+      decoys.iter().map(|decoy| Decoys::fee_weight(&decoy.offsets)).collect::<Vec<_>>();
+
+    // Deterministically calculate tx weight and fee
+    let (weight, fee) =
+      calculate_weight_and_fee(protocol, &decoy_weights, inputs.len(), outputs, extra, fee_rate);
 
     // The actual limit is half the block size, and for the minimum block size of 300k, that'd be
     // 150k
     // wallet2 will only create transactions up to 100k bytes however
     const MAX_TX_SIZE: usize = 100_000;
-
-    // This uses the weight (estimated_tx_size) despite the BP clawback
-    // The clawback *increases* the weight, so this will over-estimate, yet it's still safe
-    if estimated_tx_size >= MAX_TX_SIZE {
+    if weight >= MAX_TX_SIZE {
       Err(TransactionError::TooLargeTransaction)?;
     }
 
-    // Calculate the fee.
-    let fee = fee_rate.calculate(estimated_tx_size);
-
     // Make sure we have enough funds
     let in_amount = inputs.iter().map(|input| input.commitment().amount).sum::<u64>();
-    let out_amount = payments.iter().map(|payment| payment.1).sum::<u64>() + fee;
-    if in_amount < out_amount {
-      Err(TransactionError::NotEnoughFunds(in_amount, out_amount))?;
+    if in_amount < out_amount + fee {
+      Err(TransactionError::NotEnoughFunds(in_amount, out_amount + fee))?;
     }
 
-    if outputs > MAX_OUTPUTS {
-      Err(TransactionError::TooManyOutputs)?;
-    }
+    // Sanity check we have the expected number of change outputs
+    sanity_check_change_payment(&payments, change_address.is_some());
 
-    let mut payments = payments.drain(..).map(InternalPayment::Payment).collect::<Vec<_>>();
-    if let Some(change) = change_address {
-      payments.push(InternalPayment::Change(change, in_amount - out_amount));
+    // Modify the amount of the change output
+    if change_address.is_some() {
+      let change_payment = payments.last_mut().unwrap();
+      *change_payment =
+        InternalPayment::Change(change_address.clone().unwrap(), in_amount - out_amount - fee);
     }
 
-    Ok(SignableTransaction { protocol, r_seed, inputs, payments, data, fee })
+    // Sanity check the change again after modifying
+    sanity_check_change_payment(&payments, change_address.is_some());
```

Comment:
What's the concern here?

Reply:
That someone will accidentally break this invariant when modifying the code in the future. I could make the code cleaner to avoid any value in adding this assert here, but figured this is solid to ship and it's worth me moving on